### PR TITLE
Add import handling for external classes

### PR DESCRIPTION
**Pull request by tvillaren**
# Why

This PR adds support for imports.

The following input: 
```ts
// source.ts
import { Person } from "@3rdparty/person";

export interface Hero {
  name: string;
  realPerson: Person;
}
```

will output

```ts
//output.ts

const personSchema = z.any();

export const heroSchema = z.object({
  name: z.string(),
  realPerson: personSchema,
});
```

### CODE REVIEW

**Comment by tvillaren** [comment]
Blocked by https://github.com/fabien0102/ts-to-zod/issues/140

**Comment by tvillaren** [comment]
Hey @fabien0102, 

Thank you for your recent update & merge. 
Following those, here is proposal for a feature addition to handle non-relative imports. I already use it today (the only missing bit to ship it was the validation issue #140). 

I acknowledge that this feature is a bit opinionated as it takes any non-relative import as not being a zod schema. 
Happy to exchange about it here. 

As a follow-up, I see an opening to handle relative imports as zod schemas (generated or not) and may open up the door to multi-file generation support eventually.

**Comment by dan-blank** [comment]
Thanks for the PR!

> I acknowledge that this feature is a bit opinionated as it takes any non-relative import as not being a zod schema. Happy to exchange about it here.

I think this is a sensible default. One option came to my mind:

Use a comment to tell ts-to-zod that the import is assumed to be a zod schema.

```TypeScript
// source.ts

import { Person } from "@3rdparty/person";

// ts-to-zod ImportIsZodSchema Person

export interface Hero {
  name: string;
  realPerson: Person;
}
```

Its semantics would be:
- Once we parsed `// ts-to-zod ImportsIsZodSchema Person`, we assume as every use site of Person that Person is a zod schema
- Does not work for `* ` in imports. When encountering `*`, log an error ("Only works for explicit imports") and use the default (= assume that the import is not a zod schema)
- Does not work for `as` in imports

> It works for non-relative modules only, based on the principle that we might want to handle relative modules using the ts-to-zod conversion.

I don't understand the reasoning, could you explain/clarify in your OP? :)

**Comment by tvillaren** [comment]
The rationale behind using only non-relative imports for this feature was mostly based on the rationale that 3rd party imports had very little chance to be Zod Schema.

Relative imports, on the other hand, have maybe a higher chance to be a class / type / interface that is also transformed somewhere in the codebase into a zodSchema.
👉 For those, there are different handling strategies (like using JSTags-style comments) to tell the transpiler we have a zodSchema for this somewhere
👉 And I had as a future step to maybe fiddle with multiple file ts-to-zod (that would be based on relative imports obviously)

👉 **In the meantime, it's quite easy to handle any imports as 3rd party ones and have specific JSTags comments in the future to handle zod schema (generated or referenced manually)**

**Comment by tvillaren** [comment]
Hello @dan-blank, @fabien0102 👋 

I updated this PR to allow any type of imports (non-relative *and* relative modules alltogether). 
For now, it assumes that imports are to be used as `z.instanceof()`, I'm going to work on another PR to add JSTags for import statement to allow specifying when we reference a zod schema.

WDYT?

**Comment by anthony-dandrea** [comment]
Looks good here as well! Ran into the same small issue mentioned in your other [PR](https://github.com/fabien0102/ts-to-zod/pull/148#issuecomment-1664231249).

**Comment by tvillaren** [comment]
Hello @fabien0102 , @schiller-manuel 

After much discussions and thoughts, I reworked this Pull Request to support `imports` and generate a `z.any()` for all type references coming from an import. 

It was previously using `z.instanceof(ImportName)` but it was an overlook from my side of the [official documentation](https://zod.dev/?id=instanceof) where `instanceof` is meant to be used only for imported `Class` (not types or interfaces).

Additionally, I had to update the `fixOptionalAny` bit that was added a couple of months ago to support type references as well. 

But now this supports simply the imports, and I will rebase https://github.com/fabien0102/ts-to-zod/pull/148 to support imports of Zod Schemas as well.

File: src/core/generate.ts [hunk 1]
```diff
@@ -82,13 +87,24 @@ export function generate({
   // declare a map to store the interface name and its corresponding zod schema
   const typeNameMapping = new Map<string, TypeNode>();
 
+  // handling existing import statements
+  const importNamesAvailable = new Set<string>();
+  // const importNodes: ts.ImportDeclaration[] = [];
```
Comment (schiller-manuel) [inline src/core/generate.ts:92]: can this be removed?

File: src/core/generate.ts [hunk 2]
```diff
@@ -82,13 +87,24 @@ export function generate({
...
   const typesNeedToBeExtracted = new Set<string>();
 
   const typeNameMapBuilder = (node: ts.Node) => {
     if (isTypeNode(node)) {
       typeNameMapping.set(node.name.text, node);
     }
+    if (ts.isImportDeclaration(node) && node.importClause) {
+      const imports = getImportIdentifiers(node);
+      imports.forEach((i) => importNamesAvailable.add(i));
+      // importNodes.push(node);
```
Comment (schiller-manuel) [inline src/core/generate.ts:104]: can this be removed?

**Review by schiller-manuel — CHANGES_REQUESTED** [review]
looks good, would love to get this and #148 merged asap.
can you clean this up (see my comments) please?

**Comment by tvillaren** [comment]
Hey @schiller-manuel, 

Thank you for you review, I just cleaned the two commented lines. 
I'm working on https://github.com/fabien0102/ts-to-zod/pull/148 and will push an update ASAP.

Thanks.

**Pull request merged**